docs/local-lxd/explanation/charm-architecture.md

```diff
@@ -0,0 +1,132 @@
+would also be directed to the HTTP(s) proxy, unlike when using aproxy.
+
+### denylist
+
+The nftables on the Juju machine are configured to deny traffic from the runner virtual machine to IPs on the [`denylist` configuration](https://charmhub.io/github-runner/configure#denylist). The runner will always have access to essential services such as DHCP and DNS, regardless of the denylist configuration.
+
+## GitHub API usage
+
+The charm requires a GitHub personal access token for the [`token` configuration](https://charmhub.io/github-runner/configure#token). This token is used for:
+
+- Requesting self-hosted runner registration tokens
+- Requesting self-hosted runner removal tokens
+- Requesting a list of runner applications
+- Requesting a list of self-hosted runners configured in an organization or repository
+- Deleting self-hosted runners
+
+The token is also passed to [repo-policy-compliance](https://github.com/canonical/repo-policy-compliance) to access GitHub API for the service.
+
+Note that the GitHub API uses a [rate-limiting mechanism](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28). When this is reached, the charm may not be able to perform the necessary operations and may go into
+BlockedStatus. The charm will automatically recover from this state once the rate limit is reset, but using a different token with a higher rate limit may be a better solution depending on your deployment requirements.
+
+## GitHub repository setting check
+
+The [repo-policy-compliance](https://github.com/canonical/repo-policy-compliance) is a [Flask application](https://flask.palletsprojects.com/) hosted on [Gunicorn](https://gunicorn.org/) that provides a RESTful HTTP API to check the settings of GitHub repositories. This ensures the GitHub repository settings do not allow the execution of code not reviewed by maintainers on the self-hosted runners.
+
+Using the [pre-job script](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/running-scripts-before-or-after-a-job#about-pre--and-post-job-scripts), the self-hosted runners call the Python web service to check if the GitHub repository settings for the job are compliant. If not compliant, it will output an error message and force stop the runner to prevent code from being executed.
+
+## Dependencies management
+
+Upon installing or upgrading the charm, the kernel will be upgraded, and the Juju machine will be restarted if needed.
+
+The charm installs the following dependencies:
+
+- For running repo-policy-compliance
+  - gunicorn
+- For firewall to prevent runners from accessing web service on the denylist
+  - nftables
+- For virtualization and virtual machine management
+  - lxd
+  - cpu-checker
+  - libvirt-clients
+  - libvirt-daemon-driver-qemu
+  - apparmor-utils
+
+These dependencies can be regularly updated using the [landscape-client charm](https://charmhub.io/landscape-client).
```

Comment:
To make the style checker happy, you may want to use code styling like `landscape-client`